onnxruntime/contrib_ops/rocm/diffusion/group_norm_tunable_op.h

```diff
@@ -0,0 +1,204 @@
+  grid.z = params->n;
+
+  groupNormNHWCSumKernel<T, ThreadsPerBlock, VecSize>
+      <<<grid, ThreadsPerBlock, 0, params->stream>>>(
+          params->src, params->redBuffer, params->cPerBlock, params->hwPerBlock,
+          params->hw, params->hwc, params->c, params->cPerGroup, params->groups, params->groupsPerBlock);
+  return HIP_CALL(hipGetLastError());
+}
+
+template <typename T>
+void groupNormNHWCScale(const GroupNormNHWCParams<T>* params) {
+  // Make sure the dimensions are aligned with what we expect.
+  ORT_ENFORCE(params->c % params->cPerBlock == 0);
+  // Make sure a group does not span multiple blocks.
+  ORT_ENFORCE(params->cPerBlock % params->cPerGroup == 0);
+
+  dim3 grid;
+
+  // The number of blocks to compute all the channels.
+  grid.x = params->c / params->cPerBlock;
+  // The number of blocks to compute all the activations in a given instance.
+  grid.y = CeilDiv(params->hw, params->hwPerBlock);
+  // The number of instances.
+  grid.z = params->n;
+
+#define LAUNCH_GROUPNORM_SCALE(ThreadsPerBlock, VecSize)                                                               \
+  groupNormNHWCScaleKernel<T, ThreadsPerBlock, VecSize>                                                                \
+      <<<grid, ThreadsPerBlock, 0, params->stream>>>(params->dst, params->src, params->gamma, params->beta,            \
+                                                     params->redBuffer, params->epsilon, params->c, params->cPerBlock, \
+                                                     params->cPerGroup, params->groups, params->hwc, params->invHWC,   \
+                                                     params->hw, params->hwPerBlock, params->withSwish);               \
+  break;
+
+  switch (params->cPerBlock) {
+    case 320:
+      LAUNCH_GROUPNORM_SCALE(256, 2)
+    case 480:
+      LAUNCH_GROUPNORM_SCALE(256, 2)
+    case 256:
+      LAUNCH_GROUPNORM_SCALE(128, 2)
+    case 128:
+      LAUNCH_GROUPNORM_SCALE(64, 2)
+    default:
+      ORT_NOT_IMPLEMENTED("Not implemented");
+  }
+}
+
+template <typename T, int ThreadsPerBlock, int VecSize>
+Status GroupNormNHWCScaleOp(const GroupNormNHWCParams<T>* params) {
+  dim3 grid;
+  grid.x = params->c / params->cPerBlock;
+  grid.y = CeilDiv(params->hw, params->hwPerBlock);
+  grid.z = params->n;
+
+  groupNormNHWCScaleKernel<T, ThreadsPerBlock, VecSize>
+      <<<grid, ThreadsPerBlock, 0, params->stream>>>(
+          params->dst, params->src, params->gamma, params->beta, params->redBuffer, params->epsilon, params->c, params->cPerBlock,
+          params->cPerGroup, params->groups, params->hwc, params->invHWC, params->hw, params->hwPerBlock, params->withSwish);
+  return HIP_CALL(hipGetLastError());
+}
+
+template <typename T, int ThreadsPerBlock, int VecSize>
+class GroupNormNHWCOp {
+ public:
+  Status operator()(const GroupNormNHWCParams<T>* params) {
+    DUMP_TENSOR_INIT();
+    DUMP_TENSOR("input", params->src, params->n, params->c, params->h * params->w);
```

Comment:
> DUMP_TENSOR

Is this leftover debugging code?

Reply:
I just keep it the same like CUDA EP. If we do not define `DEBUG_GENERATION`, the debugging function will not be enabled. And I think ROCm's dump tensor has not been updated yet and needs to be supported by another PR.